storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -247,51 +263,110 @@ public boolean containsBlock(Bytes32 blockRoot) {
 
   @Override
   public BeaconState getBlockState(Bytes32 blockRoot) {
+    return getOrGenerateBlockState(blockRoot).orElse(null);
+  }
+
+  @Override
+  public BeaconState getCheckpointState(Checkpoint checkpoint) {
     readLock.lock();
     try {
-      return block_states.get(blockRoot);
+      return checkpoint_states.get(checkpoint);
     } finally {
       readLock.unlock();
     }
   }
 
   @Override
-  public boolean containsBlockState(Bytes32 blockRoot) {
+  public boolean containsCheckpointState(Checkpoint checkpoint) {
     readLock.lock();
     try {
-      return block_states.containsKey(blockRoot);
+      return checkpoint_states.containsKey(checkpoint);
     } finally {
       readLock.unlock();
     }
   }
 
   @Override
-  public BeaconState getCheckpointState(Checkpoint checkpoint) {
+  public Set<UnsignedLong> getVotedValidatorIndices() {
     readLock.lock();
     try {
-      return checkpoint_states.get(checkpoint);
+      return votes.keySet();
     } finally {
       readLock.unlock();
     }
   }
 
-  @Override
-  public boolean containsCheckpointState(Checkpoint checkpoint) {
+  private Optional<BeaconState> getOrGenerateBlockState(final Bytes32 blockRoot) {
+    Optional<BeaconState> state = getState(blockRoot);
+    if (state.isPresent()) {
+      return state;
+    }
+    final SignedBeaconBlock blockForState = getSignedBlock(blockRoot);
+    if (blockForState == null) {
+      // If we don't have the corresponding block, we can't possibly regenerate the state
+      return Optional.empty();
+    }
+
+    // Accumulate blocks until we find our base state to build from
+    SignedBlockAndState baseBlock = null;
+    final Map<Bytes32, SignedBeaconBlock> blocks = new HashMap<>();
+    SignedBeaconBlock block = blockForState;
+    while (block != null) {
+      final Optional<BeaconState> blockState = getState(block.getRoot());
+      if (blockState.isPresent()) {
+        // We found a base state
+        baseBlock = new SignedBlockAndState(block, blockState.get());
+        break;
+      }
+      blocks.put(block.getRoot(), block);
+      block = getSignedBlock(block.getParent_root());
+    }
+
+    if (baseBlock == null) {
+      // If we haven't found a base state yet, we must have walked back to the latest finalized
+      // block, check here for the base state
+      final SignedBlockAndState finalizedBlock = getLatestFinalizedBlockAndState();
+      if (!blocks.keySet().contains(finalizedBlock.getRoot())) {
+        // We must have finalized a new block while processing and moved past our target root
+        return Optional.empty();
+      }
+      baseBlock = finalizedBlock;
+    }
+
+    // Regenerate state
+    final StateGenerator stateGenerator = new StateGenerator();
+    final Map<Bytes32, BeaconState> regeneratedStates =
+        stateGenerator.produceStatesForBlocks(
+            baseBlock.getRoot(), baseBlock.getState(), blocks.values());
```

Comment:
For a follow up ticket - we should ensure we can restart in the middle of a long period of non-finalisation too.  I think currently we'd regenerate all states and need to hold them in memory which would cause issues.

Reply:
> I'd be slightly tempted to gather up the blocks to apply inside a read lock so we know the finalised state won't change, then once we have them release the lock and do the state transitions to generate the state

We could do that, but when we come back around to block pruning, that could cause issues.  So, I think I'll leave it for now ...

Reply:
Yeah fair point.

Reply:
Reworked the `StateGenerator` so that we only ever process one state at a time and limit the number of states we keep in memory.